### PR TITLE
Support opaque (and black) particles

### LINKED ISSUE #833: Add support for opaque particles
(discussed here https://github.com/overte-org/overte/issues/776)

Our particles are always transparent with additive blending (due to sorting complexities).  As a result, you can never make black particles.  We should support opaque particles by auto-detecting if the particle texture is fully opaque or uses alpha cutout.  In the future, we should support directly specifying the alpha blend mode (likely via the existing properties on material entities).

### PR DESCRIPTION
**Pull request by HifiExperiments**
Closes #833 

Particles can render opaque if their texture has no alpha or is detected as "alpha mask" (e.g. https://pngimg.com/d/grass_PNG4922.png).  If any of their alpha properties (start, middle, finish) are < 1, or their alpha spread is > 0, or they have alpha pulsing, they'll go back to transparent (+ additive blending).  At a later date, we could support overriding the automatic detection via material entities.

Screenshot 2024-03-01 130300

### CODE REVIEW

File: libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp [hunk 1]
```diff
@@ -138,26 +150,25 @@ void ParticleEffectEntityRenderer::doRenderUpdateAsynchronousTyped(const TypedEn
     _uniformBuffer.edit<ParticleUniforms>() = particleUniforms;
 }
 
-ItemKey ParticleEffectEntityRenderer::getKey() {
```
Comment (HifiExperiments) [inline libraries/entities-renderer/src/RenderableParticleEffectEntityItem.cpp:141]: this is now identical to `EntityRenderer::getKey()` so it's not needed.

(although I might need to add it back when I do GPU particles but I'll deal with that merge conflict later)

**Comment by SilverfishVR** [comment]
It's beautiful
Screenshot 2024-03-02 18 32 32

And everything seems to work as advertised 👍

**Comment by vegaslon** [comment]
Light highlights seems to show up as boxes on this pr.

**Comment by vegaslon** [comment]
overte-snap-by--on-2024-03-02_19-06-47

interesting side effect is they cast shadows

**Comment by vegaslon** [comment]
if all alpha properties are set to 1 are they suppose to render as black and have no color? this is on high setting for the graphics on a amd
overte-snap-by-vegaslon-on-2024-03-03_11-30-18

**Comment by HifiExperiments** [comment]
@vegaslon ah AMD has given us this problem in the past.  I don't have an AMD card to test but I've pushed a potential fix.

**Comment by vegaslon** [comment]
that worked to restore color to the particles, side effect is the particles no longer throw shadows (For how ever much that is worth).

**Comment by HifiExperiments** [comment]
that would explain this comment that I copied from polyvox: `// FIXME: custom pipelines like this don't handle shadows or renderLayers correctly` lol

renderLayers actually do seem to work...but it seems neither polyvox nor particles can cast shadows.  these are the only two things that use a specific codepath (custom pipelines).  I would vote to merge this as is and then I can fix that issue separately so as to not clutter up this PR.

I already have shadows working with GPU particles<sup>TM</sup> though 😄

**Comment by vegaslon** [comment]
Ok this is something, When I run the build from github actions the particle cast shadows. So ya as far as particles go this pr seems to definitely do the job.

**Comment by JulianGro** [comment]
What build did you run before? You built from source yourself?

But yeah, maybe the shader cache was just being problematic. I have seen that before on Nvidia at the very least.

**Comment by gydence** [comment]
oh weird!  I built locally and couldn’t get particles or polyvox to cast shadows, which isn’t *wholly* surprising to me because the way we pick shadow pipelines has never handled this “custom pipeline” path

**Comment by vegaslon** [comment]
> What build did you run before? You built from source yourself?
> 
> But yeah, maybe the shader cache was just being problematic. I have seen that before on Nvidia at the very least.

ya I did a self build at first since the github actions had the linux builds stuck in line, so it was not willing to release the download links for the windows one.

**Comment by HifiExperiments** [comment]
just to be sure, I went ahead and fixed custom pipeline shadows and render layers.  this only affects particles and polyvox

to test, opaque particles and polyvox should be able to cast shadows and render correctly in all renderLayers.  transparent particles should work like they used to

**Comment by vegaslon** [comment]
Everything continues to appear in order, Particles and Polyvox apear to both cast shadows now, Opaque textures appear to work and transparent particles continue to function.

**Review by ksuprynowicz — APPROVED** [review]
Everthing looks fine :)

**Pull request merged**